tf2/include/tf2/time.h

```diff
@@ -52,7 +53,24 @@ namespace tf2
     return std::chrono::system_clock::now();
   }
 
-  inline double durationToSec(const tf2::Duration & input){
+  inline Duration durationFromSec(double t_sec)
+  {
+    uint32_t sec, nsec;
+    sec = (uint32_t)floor(t_sec);
+    nsec = (uint32_t)std::round((t_sec-sec) * 1e9);
```

Comment:
Have you tried this function with `t_sec` < 0? I think you might want `std::trunc` rather than `std::floor`, and that makes me think that the nanoseconds logic might not be perfect either.

Maybe you could use this function's "possible implementation":

http://en.cppreference.com/w/cpp/chrono/duration/round

It is not available until C++17, but it's possible implementation might work in more cases than this.